src/sctools/test/test_metrics.py

```diff
@@ -0,0 +1,374 @@
+    assert observed == expected_value
+
+
+@pytest.mark.parametrize('metrics, expected_value', [
+    (_gene_metrics, 0),  # todo null case
+    (_cell_metrics, 19),
+])
+def test_reads_mapped_utr(metrics, expected_value):
+    """Test that each metric identifies the number of reads mapped to a UTR (XF=='UTR')"""
+    observed = metrics['reads_mapped_utr'].sum()
+    assert observed == expected_value
+
+
+@pytest.mark.parametrize('metrics, expected_value', [
+    (_gene_metrics, 300),  # todo need to include at least 1 multi-mapper
+    (_cell_metrics, 656),
+])
+def test_reads_mapped_uniquely(metrics, expected_value):
+    """Uniquely mapping reads will be tagged with NH==1"""
+    observed = metrics['reads_mapped_uniquely'].sum()
+    assert observed == expected_value
+
+
+@pytest.mark.parametrize('metrics, expected_value', [
+    (_gene_metrics, 90),
+    (_cell_metrics, 107),
+])
+def test_duplicate_records(metrics, expected_value):
+    """Duplicate records are identified by the 1024 bit being set in the sam flag"""
+    observed = metrics['duplicate_reads'].sum()
+    assert observed == expected_value
+
+
+@pytest.mark.parametrize('metrics, expected_value', [
+    (_gene_metrics, 29),
+    (_cell_metrics, 2),
+])
+def test_spliced_reads(metrics, expected_value):
+    """
+    This pipeline defines spliced reads as containing an N segment of any length in the cigar string
+    """
+    observed = metrics['spliced_reads'].sum()
+    assert observed == expected_value
+
+
+# todo failing
+@pytest.mark.parametrize('metrics', [_gene_metrics, _cell_metrics])
+def test_relationship_of_duplicates_and_fragments(metrics):
+    """
+    We expect the number of duplicates and fragments to add up to the total number of reads. The
+    rationale is that any read that is not a duplicate should be a distinct fragment, under our
+    definitions.
+
+    This fails because of (1) N-base and 2-base cell barcode correction errors and (2)
+    fragment calculationes currently do not account for soft clipping. Fixing these will cause this
+    test to pass
+    """
+    dup_and_fragments = metrics['duplicate_reads'].sum() + metrics['n_fragments'].sum()
+    reads = metrics['n_reads'].sum()
+    assert reads == dup_and_fragments
+
+
+@pytest.mark.parametrize('metrics', [_gene_metrics, _cell_metrics])
+def test_fragments_number_is_greater_than_molecule_number(metrics):
+    """
+    There should always be more fragments than molecules, as the minimum definition of a molecule is
+    a fragment covered by a single read
+    """
+    assert np.all(metrics['n_molecules'] >= 1)
+    assert np.all(metrics['n_fragments'] >= 1)
+    assert np.all(metrics['n_fragments'] >= metrics['n_molecules'])
+
+
+@pytest.mark.parametrize('metrics, key, expected_value', [
+    (_cell_metrics, 'molecule_barcode_fraction_bases_above_30_mean',
+     np.array([1.0000, 0.9500, 1.0000, 1.0000, 0.9778, 1.0000, 1.0000, 1.0000, 0.9833, 1.0000,
```

Comment:
Are these just random inputs you created for test purposes?

Reply:
No, they are not. We examined the test data in the notebooks enclosed in this PR. We used different tools to arrive at what we expect the metric outputs to be and verify that the metric iterators defined for the production pipelines match our expectations.

Reply:
Ok, thanks for the clarification! Can you add this in as a comment to clarify where this test data is coming from?